File: apps/api/app/core/currency.py
"""Currencies the workspace can keep its books in.

Most are ISO 4217 and format themselves through the standard library. Toman is
the exception worth the code: it is what Iranians actually quote prices in, it
has no ISO code, and it is worth ten rial — so it is carried here explicitly
rather than approximated with IRR.
"""

from dataclasses import dataclass


@dataclass(frozen=True)
class Currency:
    code: str
    name_en: str
    name_fa: str
    symbol_en: str
    symbol_fa: str
    decimals: int
    iso: bool = True          # False when Intl/babel will not know the code
    symbol_position: str = "before"   # before | after
    #: Value of one unit in the ISO currency it derives from, when not ISO.
    base_code: str | None = None
    base_rate: float = 1.0


CURRENCIES: list[Currency] = [
    Currency("IRT", "Toman", "تومان", "IRT", "تومان", 0,
             iso=False, symbol_position="after", base_code="IRR", base_rate=10.0),
    Currency("IRR", "Iranian rial", "ریال ایران", "﷼", "ریال", 0,
             symbol_position="after"),
    Currency("USD", "US dollar", "دلار آمریکا", "$", "$", 2),
    Currency("EUR", "Euro", "یورو", "€", "€", 2),
    Currency("GBP", "Pound sterling", "پوند", "£", "£", 2),
    Currency("AED", "UAE dirham", "درهم امارات", "AED", "درهم", 2,
             symbol_position="after"),
    Currency("TRY", "Turkish lira", "لیر ترکیه", "₺", "₺", 2),
    Currency("IQD", "Iraqi dinar", "دینار عراق", "IQD", "دینار", 0,
             symbol_position="after"),
    Currency("RUB", "Russian rouble", "روبل روسیه", "₽", "₽", 2),
    Currency("CNY", "Chinese yuan", "یوان چین", "¥", "¥", 2),
    Currency("INR", "Indian rupee", "روپیه هند", "₹", "₹", 2),
    Currency("CAD", "Canadian dollar", "دلار کانادا", "CA$", "CA$", 2),
    Currency("AUD", "Australian dollar", "دلار استرالیا", "A$", "A$", 2),
    Currency("CHF", "Swiss franc", "فرانک سوئیس", "CHF", "فرانک", 2),
    Currency("JPY", "Japanese yen", "ین ژاپن", "¥", "¥", 0),
    Currency("SEK", "Swedish krona", "کرون سوئد", "kr", "کرون", 2,
             symbol_position="after"),
]

BY_CODE = {c.code: c for c in CURRENCIES}
DEFAULT = "USD"


def get(code: str | None) -> Currency:
    return BY_CODE.get((code or DEFAULT).upper(), BY_CODE[DEFAULT])


def is_supported(code: str | None) -> bool:
    return bool(code) and code.upper() in BY_CODE


def as_dicts() -> list[dict]:
    return [
        {
            "code": c.code,
            "name_en": c.name_en,
            "name_fa": c.name_fa,
            "symbol_en": c.symbol_en,
            "symbol_fa": c.symbol_fa,
            "decimals": c.decimals,
            "iso": c.iso,
            "symbol_position": c.symbol_position,
            "base_code": c.base_code,
            "base_rate": c.base_rate,
        }
        for c in CURRENCIES
    ]

File: apps/web/lib/currency.ts
/**
 * Currencies the workspace can keep its books in.
 *
 * Most are ISO 4217 and go through Intl. Toman is the one worth the code: it is
 * what Iranians actually quote prices in, it has no ISO code — so Intl throws on
 * it — and it is worth ten rial. It is therefore formatted by hand, with the
 * symbol after the amount the way Persian writes it.
 */

export type Currency = {
  code: string;
  name_en: string;
  name_fa: string;
  symbol_en: string;
  symbol_fa: string;
  decimals: number;
  iso: boolean;
  symbol_position: "before" | "after";
  base_code?: string | null;
  base_rate?: number;
};

export const CURRENCIES: Currency[] = [
  { code: "IRT", name_en: "Toman", name_fa: "تومان", symbol_en: "IRT", symbol_fa: "تومان",
    decimals: 0, iso: false, symbol_position: "after", base_code: "IRR", base_rate: 10 },
  { code: "IRR", name_en: "Iranian rial", name_fa: "ریال ایران", symbol_en: "﷼",
    symbol_fa: "ریال", decimals: 0, iso: true, symbol_position: "after" },
  { code: "USD", name_en: "US dollar", name_fa: "دلار آمریکا", symbol_en: "$",
    symbol_fa: "$", decimals: 2, iso: true, symbol_position: "before" },
  { code: "EUR", name_en: "Euro", name_fa: "یورو", symbol_en: "€", symbol_fa: "€",
    decimals: 2, iso: true, symbol_position: "before" },
  { code: "GBP", name_en: "Pound sterling", name_fa: "پوند", symbol_en: "£", symbol_fa: "£",
    decimals: 2, iso: true, symbol_position: "before" },
  { code: "AED", name_en: "UAE dirham", name_fa: "درهم امارات", symbol_en: "AED",
    symbol_fa: "درهم", decimals: 2, iso: true, symbol_position: "after" },
  { code: "TRY", name_en: "Turkish lira", name_fa: "لیر ترکیه", symbol_en: "₺",
    symbol_fa: "₺", decimals: 2, iso: true, symbol_position: "before" },
  { code: "IQD", name_en: "Iraqi dinar", name_fa: "دینار عراق", symbol_en: "IQD",
    symbol_fa: "دینار", decimals: 0, iso: true, symbol_position: "after" },
  { code: "RUB", name_en: "Russian rouble", name_fa: "روبل روسیه", symbol_en: "₽",
    symbol_fa: "₽", decimals: 2, iso: true, symbol_position: "before" },
  { code: "CNY", name_en: "Chinese yuan", name_fa: "یوان چین", symbol_en: "¥",
    symbol_fa: "¥", decimals: 2, iso: true, symbol_position: "before" },
  { code: "INR", name_en: "Indian rupee", name_fa: "روپیه هند", symbol_en: "₹",
    symbol_fa: "₹", decimals: 2, iso: true, symbol_position: "before" },
  { code: "CAD", name_en: "Canadian dollar", name_fa: "دلار کانادا", symbol_en: "CA$",
    symbol_fa: "CA$", decimals: 2, iso: true, symbol_position: "before" },
  { code: "AUD", name_en: "Australian dollar", name_fa: "دلار استرالیا", symbol_en: "A$",
    symbol_fa: "A$", decimals: 2, iso: true, symbol_position: "before" },
  { code: "CHF", name_en: "Swiss franc", name_fa: "فرانک سوئیس", symbol_en: "CHF",
    symbol_fa: "فرانک", decimals: 2, iso: true, symbol_position: "before" },
  { code: "JPY", name_en: "Japanese yen", name_fa: "ین ژاپن", symbol_en: "¥",
    symbol_fa: "¥", decimals: 0, iso: true, symbol_position: "before" },
  { code: "SEK", name_en: "Swedish krona", name_fa: "کرون سوئد", symbol_en: "kr",
    symbol_fa: "کرون", decimals: 2, iso: true, symbol_position: "after" },
];

const BY_CODE = new Map(CURRENCIES.map((c) => [c.code, c]));

export function currencyOf(code?: string | null): Currency {
  return BY_CODE.get((code ?? "USD").toUpperCase()) ?? BY_CODE.get("USD")!;
}

export function currencyName(code: string, locale: string) {
  const currency = currencyOf(code);
  return locale === "fa" ? currency.name_fa : currency.name_en;
}

export function currencySymbol(code: string, locale: string) {
  const currency = currencyOf(code);
  return locale === "fa" ? currency.symbol_fa : currency.symbol_en;
}
